README.md

```diff
@@ -1,138 +1,664 @@
+
+Using a prefix lets us restrict in how decorators are defined and used. The ordinary JavaScript lexical variable scope permits various kinds of dynamism, e.g., through `with` statements, the global object, and `var` declarations leaking out of eval. By using a prefix, we can define those sources of variability away.
+
+We've found that it's intuitive to have `@` as part of the name of decorators: Most documentation for decorators in practice today treated `@` as if it were part of the name, rather than the syntax for invoking decorators.
+
+As a bonus: The separate namespace for the new, static decorators proposal should also help the transition path from previous transpiler-based decorators proposals: It's easy for tooling to see whether you're referring to a static decorator or not, just by what names it can see in scope.
+
+### Could we support decorating objects, parameters, blocks, functions, etc?
+
+Yes! Once we have validated this core approach, the authors of this proposal plan to come back and make proposals for more kinds of decorators. See [NEXTBUILTINS.md](./NEXTBUILTINS.md#applying-built-in-decorators-to-other-syntactic-forms).
+
+### Will decorators let you access private fields and methods?
+
+The champion group feels very strongly that decorator access to private fields and methods in various different ways is very important, and we're trying to provide it as soon as possible. This proposal does not include any built-in decorators that would provide the primitives to access private fields or methods (beyond wrapping them). We hope to provide this capability with future built-in decorators. See [NEXTBUILTINS.md](./NEXTBUILTINS.md#expose). The focus of this proposal is on the *infrastructure* for built-in and user-defined decorators, and a minimum of functionality is provided.
+
+### When are decorators evaluated?
+
+The arguments to a decorator are evaluated inline with class evaluation, just like computed property names.
+
+The built-in decorators take callbacks as arguments, which are scheduled to run later at different times:
+- The `@wrap` wrapping function is executed while setting up the class.
+- The `@register` callback is executed after the class is created.
+- The `@initialize` callback is called just after executing the class initializer (normally, in the constructor).
+
+Whenever there are multiple callbacks, they are executed from "top to bottom, inside to out", regardless of the type or placement of class element. This goes for all three built-in decorators. Here's an example based on `@register`:
+
+```js
+decorator @log(msg) { @register(k => { console.log(msg); return k }) }
+
+@log("a") @log("b")
+class C {
+  @log("d") @log("e") method() { }
+
+  @log("f") @log("g") static prop;
+}
+```
+
+This example logs `"e"`, `"d"`, `"g"`, `"f"`, `"b"`, `"a"`.
```

Comment:
One of the most important issue I would imagine is the data transition some information with connected decorators (class -> property/method or property/method -> class). So the execution order is very important here. By default, I do the following:
```javascript
const map = new WeakMap();

const cls = ({elements, kind}) => {
  return {
    elements: [
      ...elements,
      {
        initializer() { // executes BEFORE all field finishers executed
          map.set(this, someData);
        }
        kind: 'initializer',
        placement: 'static',
      } 
    ],
    kind,
  }
};

const property = (descriptor) => {
  let $data;

  return {
    ...descriptor,
    {
      finisher() { // executes AFTER all class initializer executed
        $data = map.get(this);
      },
      kind: 'initializer',
      placement: 'static',
    }
  }
}
```
However, with the order PR proposes it looks impossible to do this kind of trick for transfer data from the class decorator to a property/method one because property/method `@register` will be executed earlier than class one. 

I wonder what can we do here? Set some kind of priority?

Reply:
Ok, I solved my issue. I can put things that should be initialized before the `@register` into a `@wrap` that returns the same class instance it receives but produces side effects:
```javascript
const map = new WeakMap();

decorator @cls {
  @wrap(target => {
    map.set(target, someData);
    return target;
  });
}

decorator @property {
  let $data;

  @register(target => {
    $data = map.get(target);
  })
}
```